plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/eap/OpenShiftEapPublishController.java

```diff
@@ -10,29 +10,78 @@
  ******************************************************************************/
 package org.jboss.tools.openshift.core.server.behavior.eap;
 
+import java.io.File;
+import java.util.Arrays;
+
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.wst.server.core.IModule;
 import org.eclipse.wst.server.core.IServer;
+import org.jboss.ide.eclipse.as.core.server.internal.v7.DeploymentMarkerUtils;
 import org.jboss.ide.eclipse.as.wtp.core.server.behavior.ISubsystemController;
+import org.jboss.tools.common.util.FileUtils;
 import org.jboss.tools.openshift.core.server.behavior.OpenShiftPublishController;
 
 public class OpenShiftEapPublishController extends OpenShiftPublishController implements ISubsystemController {
 
-	public OpenShiftEapPublishController() {
-		super();
+	private static final long WAIT_FOR_UNDEPLOYED_TIMEOUT = 20l * 1000l; // 10s
+	private static final long UNDEPLOYED_CHECKS = 10;
+
+	@Override
+	public int publishModule(int kind, int deltaKind, IModule[] module, IProgressMonitor monitor) throws CoreException {
+		return super.publishModule(IServer.PUBLISH_CLEAN, deltaKind, module, monitor);
 	}
 
 	@Override
-	public void publishFinish(IProgressMonitor monitor) throws CoreException {
-		super.publishFinish(monitor);
+	protected File deleteDeploymentFile(File moduleDeployment, IProgressMonitor monitor) throws CoreException {
+		try {
+			File moduleDeploymentFile = super.deleteDeploymentFile(moduleDeployment, monitor);
+			if (moduleDeploymentFile != null) {
+				triggerUndeploy(moduleDeploymentFile);
+				// allow server to undeploy first by syncing right after remove
+				syncDirectoryToPods(monitor);
+				// wait for undeployed marker to show up
+				waitForUndeployed(moduleDeployment, monitor);
+			}
+			return moduleDeploymentFile;
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			return null;
+		}
+	}
 
-		deleteDoDeployMarkers(getDeploymentsRootFolder());
+	private void triggerUndeploy(File moduleDeployment) {
+		deleteFiles(new File(moduleDeployment + DeploymentMarkerUtils.DEPLOYED),
+				new File(moduleDeployment + DeploymentMarkerUtils.DO_DEPLOY),
+				new File(moduleDeployment + DeploymentMarkerUtils.DEPLOYING));
 	}
 
-	@Override
-	public int publishModule(int kind, int deltaKind, IModule[] module, IProgressMonitor monitor) throws CoreException {
-		return super.publishModule(IServer.PUBLISH_CLEAN, deltaKind, module, monitor);
+	private void deleteFiles(File... files) {
+		Arrays.stream(files).forEach(file -> {
+			if (file.exists()) {
+				FileUtils.remove(file);
+			}
+		});
+	}
+
+	private void waitForUndeployed(File moduleDeployment, IProgressMonitor monitor) throws CoreException, InterruptedException {
```

Comment:
I worry that if you try to delete something which was incompletely deployed or copied over but not actually deployed by the server, you may end up waiting the full timeout because you're waiting for the server to recognize the undeployment of something it never fully deployed. Not sure how you'd accurately detect this in advance to reduce the chance of long timeouts.

Reply:
@robstryker maybe detecting a .failed marker and not wait in this case would be a valuable . Other than this I dont see how we'd detect such a failed deployment.

Reply:
@robstryker looking at the code I see the ```waitForUndeployed``` looking for a .failed marker and stopping the wait if it is detected. 
```
	private void waitForUndeployed(File moduleDeployment, IProgressMonitor monitor) throws CoreException, InterruptedException {
		long start = System.currentTimeMillis();
		do {
				Thread.sleep(WAIT_FOR_UNDEPLOYED_TIMEOUT/UNDEPLOYED_CHECKS);
				syncPodsToDirectory(monitor);
		}
		while(!isUndeployed(moduleDeployment)
				&& System.currentTimeMillis() < start + WAIT_FOR_UNDEPLOYED_TIMEOUT
				&& !monitor.isCanceled());
	}

	private boolean isUndeployed(File moduleDeployment) {
		return fileExists(new File(moduleDeployment + DeploymentMarkerUtils.UNDEPLOYED),
				new File(moduleDeployment + DeploymentMarkerUtils.FAILED_DEPLOY));
	}
```
In that sense I believe that the most we can do is implemented. Right?